Misc fixes

FreeMasen opened the pull request:
This PR addresses a number of bug and/or ergonomic issues with the library.

First and foremost this begins the process of centralizing a lot of the code that was duplicated in both the request and response modules, this makes addressing the following issues easier.

- the `iter` and `serialize` APIs were only available for the "builder" path, which is a bit of an ergonomic nightmare (#3), this fixes that
- The `get_content_length` method wasn't available on both types (one was left not updated previously from the unprefixed setter)
- The `get_*`, `fill_*` and `seralize*` methods are all nearly identical

Additionally, this PR addresses some general cleanup around docs test and type checking

FreeMasen committed:
fix: response serialization when streaming from source

FreeMasen committed:
fix: Response status msg can contain spaces

FreeMasen committed:
fix: response doc fixes

FreeMasen committed:
fix: allow headers to be duplicated on serialization

FreeMasen committed:
chore: start moving logic into shared module

FreeMasen committed:
chore: start moving logic into shared module

FreeMasen committed:
fix: content-length and content-type should only have 1 entry

FreeMasen committed:
chore: misc doc fixes

FreeMasen committed:
chore: misc type checking fixes

FreeMasen committed:
fix: sources should respect patterns

FreeMasen committed:
chore: fix tests

FreeMasen committed:
chore: fix readme examples

FreeMasen committed:
chore: fix readme examples

FreeMasen committed:
chore: ignore test dir when generating coverage

FreeMasen committed:
doc: add documentation about sources

FreeMasen committed:
fix: github pages from branch

FreeMasen committed:
chore: update test action to trigger on main

FreeMasen committed:
fix: outdated examples

FreeMasen committed:
add send logic to shared

FreeMasen committed:
update luarocks version

FreeMasen committed:
increase coverage

FreeMasen committed:
fix bad assertion

FreeMasen committed:
add chunk encoding

FreeMasen committed:
better doc comment on serialize_header

FreeMasen committed:
more type hints

FreeMasen committed:
more robust header parsing from chunks

FreeMasen committed:
restore deprecated method Request:content_length

dljsjr commented:
I haven't had a chance to dig in really deep yet, does this refactor address chunked transfer encoding for socket ltn12 sources?

FreeMasen commented:
> I haven't had a chance to dig in really deep yet, does this refactor address chunked transfer encoding for socket ltn12 sources?

Yes, the primary logic for this is [here](https://github.com/cosock/luncheon/blob/e956f08b52778fe14346a6eee39f75e42d2c5475/luncheon/shared.lua#L83-L205), with additional modifications [here](https://github.com/cosock/luncheon/blob/e956f08b52778fe14346a6eee39f75e42d2c5475/luncheon/shared.lua#L304-L310) for the `iter` implementation

FreeMasen committed:
add log dependency for warning

FreeMasen committed:
Handle chunks larger than 255 bytes
Account for extensions on length line

FreeMasen committed:
fix doc comment about serialize header argument types

FreeMasen committed:
handle trailers on incoming

FreeMasen committed:
misc test updates for chunk encoding

FreeMasen committed:
also test response chunk encoding

blueyetisoftware committed:
Strip blank lines leading the preamble

blueyetisoftware committed:
Prevent lock up for keep-alive sockets without content-length

FreeMasen committed:
more chunk encoding

FreeMasen committed:
Merge commit '324ff937c936137979e59fa3db295f543c81f4be' into blue-yeti-merge

FreeMasen committed:
Merge commit '38f62a1b31eb2c2ee2e51259caf0633243c5bb3f' into blue-yeti-merge

FreeMasen committed:
more chunk-encoding support

Commit:
test for long first line

Commit:
_send_state.sent default 0 when nil

Commit:
formatting

Commit:
more tests

FreeMasen committed:
more robust chunked encoding detection

FreeMasen committed:
more docs

FreeMasen committed:
more statuses

Commit:
accept more methods

cjswedes commented:
@FreeMasen what are you thinking that you still want to do here? Currently there are some failing tests relating to the chunked encoding. Let me know if there is anything I can take on to help get this completed.

FreeMasen commented:
> @FreeMasen what are you thinking that you still want to do here? Currently there are some failing tests relating to the chunked encoding. Let me know if there is anything I can take on to help get this completed.

It looks like the CI run passed and I am passing locally, are you seeing failures locally?

cjswedes commented:
I think I had something off with my environment. I am seeing them pass now.

cjswedes commented on luncheon/shared.lua:
```diff
@@ -0,0 +1,618 @@
+  return 1
+end
+
+return {
```
It seems like this entire shared logic module could be included into the Request/Response classes as a metatable. We dont have to make such a change right now, but I was curious if going down the route of setting up inheritance in this way was considered.

FreeMasen replied:
yeah, this comes down to me not really understanding lua's inheritance model and this was the easiest way to share that logic. I am not opposed to using a meta-table to achieve the same thing but just haven't had the energy to tackle both the parsing logic and the inheritance layout.

cjswedes replied:
https://github.com/cosock/luncheon/pull/10

dljsjr committed:
Update header parsing to handle headers that are key-only

An example is the `EXT:` header from RFC 2774: https://datatracker.ietf.org/doc/html/rfc2774#section-4.3

dljsjr committed:
Update tests and some logic around key-only headers.

cjswedes committed:
remove unnecessary Mode

cjswedes committed:
fixup! remove unnecessary Mode

cjswedes committed:
fixup! typo in Request.new()

cjswedes committed:
make request source same as response source

cjswedes committed:
fixup! remove unused _recvd field on response

cjswedes committed:
Implement metatable inheritance to remove extra functions

There is some cleanup needed, and some other functionality that could
likely be shared in ReqResp class

cjswedes committed:
Move more common header manipulation functions to ReqResp

cjswedes committed:
fixup! Move more common header manipulation functions to ReqResp

cjswedes committed:
remove unused function

cjswedes committed:
use method syntax for ReqResp

dljsjr committed:
Merge pull request #9 from dljsjr/misc-fixes

cjswedes committed:
Merge branch 'misc-fixes' into refactor-shared-code

cjswedes committed:
rename ReqResp to HttpMessage

cjswedes committed:
fixup! unnecessary local variables

cjswedes committed:
Merge pull request #10 from cosock/refactor-shared-code

Refactor shared code

cjswedes committed:
Merge branch 'main' into misc-fixes
